### PR TITLE
Clone on copy

### PR DESCRIPTION
**Pull request by nikku**
This fixes a fundamental flaw in our copy mechanism.

Until today we would carry around a mutable copy of `oldBusinessObject` and `oldDi` that would change what has been copied if the copied element changes after the copy operation:

```
Copy <Task> with name <OLD>
Change <Task.name> property to <FOO>
Paste <Task'>
<Task'.name> is <FOO> despite it having name <OLD> at the time of copy
```

Unfortunately the PR got a little bigger than intended, fixing numerous small issues _and_ addition additional test coverage to test copy + paste end-to-end.

### Contributions

* fixes the issue by cloning each element _as we copy_. As a result what is copied is frozen at copy time and not affected by future changes anymore.
* fixes that paste behaviors would execute when pasting both labels and actual shapes (https://github.com/bpmn-io/bpmn-js/pull/1707/commits/fac5f92c3dd80ad488d8b855ab8027f617e94046, verified via https://github.com/bpmn-io/bpmn-js/pull/1707/commits/076bb35d42897a277459b0dafabd9db8126a3736)
* fix bogus meta-date being attached to shapes on paste (https://github.com/bpmn-io/bpmn-js/pull/1707/commits/1f40b65f537957dfe65c7484c1ce70aba38d71fc)
* adds a bunch of test cases to verify paste works in the way intended: 
    * keeping IDs where possible, working with serialization (https://github.com/bpmn-io/bpmn-js/pull/1707/commits/076bb35d42897a277459b0dafabd9db8126a3736)
    * verify collapsed sub-process handling (https://github.com/bpmn-io/bpmn-js/pull/1707/commits/63e6e2ec11735ad9d78f5f2af80e8c4412c1d074)

----

Closes https://github.com/bpmn-io/bpmn-js/issues/798
Closes https://github.com/bpmn-io/bpmn-js/issues/1708

### CODE REVIEW

File: lib/features/modeling/behavior/BoundaryEventBehavior.js [hunk 1]
```diff
@@ -9,13 +9,11 @@ import {
...
 } from 'min-dash';
 
-var HIGH_PRIORITY = 2000;
-
 
 /**
  * BPMN specific boundary event behavior
  */
-export default function BoundaryEventBehavior(eventBus, moddle, modeling) {
+export default function BoundaryEventBehavior(eventBus, modeling) {
```
Comment (nikku) [inline lib/features/modeling/behavior/BoundaryEventBehavior.js:16]: **TODO:** Move `processRef` test case over.

File: lib/features/modeling/behavior/RootElementReferenceBehavior.js [hunk 2]
```diff
@@ -159,6 +156,8 @@ export default function RootElementReferenceBehavior(
     }
 
     setRootElement(businessObject, referencedRootElement);
+
+    delete descriptor.referencedRootElement;
```
Comment (nikku) [inline lib/features/modeling/behavior/RootElementReferenceBehavior.js:160]: **TODO:** add a test case for this

File: test/spec/ModelerSpec.js [hunk 3]
```diff
@@ -708,6 +715,133 @@ describe('Modeler', function() {
...
+            'bpmn:Signal',
+            'bpmn:Message'
+          ];
+
+          var globalElements = definitions.get('rootElements').filter(function(element) {
+            return isAny(element, globals);
+          });
+
+          // expect <expectedGlobals> + auto created category value
+          expect(globalElements).to.have.length(expectedGlobals.length + 1);
```
Comment (nikku) [inline test/spec/ModelerSpec.js:819]: **TODO:** Category value no longer auto generated

File: lib/features/modeling/behavior/GroupBehavior.js [hunk 4]
```diff
@@ -157,9 +218,9 @@ export default function GroupBehavior(
     }
   });
 
-  /**
-   * create new category + value when group was created
-   */
+
+  // create new category + value when group was created
+
   this.execute('shape.create', function(event) {
```
[the remark this answers is not in the document]
Comment (nikku) [reply lib/features/modeling/behavior/GroupBehavior.js:224]: We only create category value if passed via business object now.

File: lib/features/modeling/behavior/RootElementReferenceBehavior.js (hunk 2 above)
Comment (nikku) [reply lib/features/modeling/behavior/RootElementReferenceBehavior.js:160]: Added a test case for this.

File: test/spec/ModelerSpec.js (hunk 3 above)
Comment (nikku) [reply test/spec/ModelerSpec.js]: Accounted for.

**Comment by nikku** [comment]
Updated as discussed.

Using `updateModdleProperties` did not work as it gives me now way to wire parent child relationships at the moment via the secret `$parent` property.

**Comment by nikku** [comment]
To be released as a new major version.

**Comment by philippfromme** [comment]
> Using `updateModdleProperties` did not work as it gives me now way to wire parent child relationships at the moment. We'd need at least the ability to set `$parent` as a special prop.

You mean for adding root elements? Then we'll add this ability in the future.

File: lib/features/copy-paste/BpmnCopyPaste.js [hunk 5]
```diff
@@ -59,9 +59,17 @@ export default function BpmnCopyPaste(bpmnFactory, eventBus, moddleCopy) {
     }
   });
 
-  var references;
+  var referencesKey = '__bpmn-js-refs';
```
Comment (philippfromme) [inline lib/features/copy-paste/BpmnCopyPaste.js:62]: Using the ID `__bpmn-js-refs` will break copy and paste. I know this is an edge case but this is why the cache wasn't used to store the references.

Example diagram:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:modeler="http://camunda.org/schema/modeler/1.0" id="Definitions_152mxu0" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.2.0" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="8.0.0">
  <bpmn:process id="Process_11wdnxl" isExecutable="true">
    <bpmn:startEvent id="__bpmn-js-refs" name="__bpmn-js-refs">
      <bpmn:messageEventDefinition id="MessageEventDefinition_1sfp0d2" messageRef="Message_2dgisck" />
    </bpmn:startEvent>
  </bpmn:process>
  <bpmn:message id="Message_2dgisck" name="Message_1" />
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Process_11wdnxl">
      <bpmndi:BPMNShape id="Event_10uizli_di" bpmnElement="__bpmn-js-refs">
        <dc:Bounds x="179" y="79" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="160" y="122" width="74" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
```

brave_bZsXtF6dPF

**Review by philippfromme — CHANGES_REQUESTED** [review]
We should consider fixing this.

File: lib/features/copy-paste/BpmnCopyPaste.js (hunk 5 above)
Comment (nikku) [reply lib/features/copy-paste/BpmnCopyPaste.js:62]: Let's choose an identifier that starts with a non-ID prefix then, i.e. a number :wink:

Comment (nikku) [reply lib/features/copy-paste/BpmnCopyPaste.js:62]: Or this: https://www.compart.com/en/unicode/U+1F697

Comment (nikku) [reply lib/features/copy-paste/BpmnCopyPaste.js]: Now using a non-QName key to do the job: https://github.com/bpmn-io/bpmn-js/pull/1707/commits/7914e5da73f47b51951876c074e149edb4b90793#diff-6ec8b9f14d0e1b571840b0b87229ca40f6974c253b87200c33d53f62a4eeaaddR62.

Comment (philippfromme) [reply lib/features/copy-paste/BpmnCopyPaste.js]: 🤯

**Review by philippfromme — APPROVED** [review]
✅

**Pull request merged**